### PR TITLE
More thorough documancy for Platform.program because it's complex and…

### PR DESCRIPTION
**Pull request by Pilatch**
… confusing, and other sources of information don't paint a clear enough picture for beginners.

I spent a long time trying to figure this out, and nothing official was helping. This is documentation from a beginner, for beginners. It spells out what's going on in more gory detail to spare others pain of integrating Elm into arbitrary JavaScript.

### CODE REVIEW

**Comment by Pilatch** [comment]
Working on making it better. I should have an update soon.

**Comment by Pilatch** [comment]
How do I get this thing to re-run the tests without making another commit?

**Comment by OvermindDL1** [comment]
> How do I get this thing to re-run the tests without making another commit?

An owner of this repo can restart it, but as a user I think the only way is to just close then re-open this pull request (not open a new one, just re-open the same old one).

**Comment by Pilatch** [comment]
Cool. I'll try that.

**Comment by Pilatch** [comment]
Neato. It's doing it.

**Comment by Pilatch** [comment]
Trying one last time just in case the test failure is intermittent.

**Comment by richardhaven** [comment]
I am a detailed documentation editor. I'm happy to review stuff

**Comment by Pilatch** [comment]
Yea, a review would be great!

File: src/Platform.elm [hunk 1]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+The second argument, `update` is a function that takes these two arguments:
+
+1. A message from the outside world, that has the information your code will act upon.
+Usually this is a `Msg` type that you specify, (a constructor), which takes a record type
+matching the data coming from an incoming port. Unpack that data with `case` ... `of`.
+
+2. The current state of your program's model, as maintained by the Elm runtime.
+
+The `update` function must return a [Tuple](Tuple)
+containing the modified form of your program's model, and a command.
```
Comment (richardhaven) [inline src/Platform.elm]: The "command" gets fed back into the update unless it is Cmd.none

File: src/Platform.elm [hunk 2]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+type Msg = Name String
+
+type alias Model = { greetings : Int }
+
+update : Msg -> Model -> ( Model, Cmd a )
+update msg model =
+    let newModel =
+            model.greetings + 1 |> Model
+
+        greeting = case msg of
```
Comment (richardhaven) [inline src/Platform.elm]: single-case is a bad code example. Use "if" to check for a single case

File: src/Platform.elm [hunk 3]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+
+1. A message from the outside world, that has the information your code will act upon.
+Usually this is a `Msg` type that you specify, (a constructor), which takes a record type
+matching the data coming from an incoming port. Unpack that data with `case` ... `of`.
+
+2. The current state of your program's model, as maintained by the Elm runtime.
+
+The `update` function must return a [Tuple](Tuple)
+containing the modified form of your program's model, and a command.
+Generate that command by passing data to an outbound port. The data type
```
Comment (richardhaven) [inline src/Platform.elm]: The command can be created by passing data to a port, but that is a specific (if common) case. Separate that from this quick overview

File: src/Platform.elm [hunk 4]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+held by this command is a determined by Elm. It probably will __not__ match any type
+you have defined in your code. Here's an example where we output a [String](String).
+
+```elm
+port greet : String -> Cmd a
+
+type Msg = Name String
+
+type alias Model = { greetings : Int }
+
```
Comment (richardhaven) [inline src/Platform.elm]: where is model = Model(0)

File: src/Platform.elm [hunk 5]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+type alias PasswordData =
+    { userId : Int, newPassword : String }
+
+port changePwPort : (PasswordData -> msg) -> Sub msg
+
+type Msg = ChangePassword PasswordData
+
+type alias Model =
+    { numChanges : Int}
+-- Assume the update function increments `numChanges`.
```
Comment (richardhaven) [inline src/Platform.elm]: where is model = Model(0) ?

File: src/Platform.elm (hunk 2 above)
Comment (Pilatch) [inline src/Platform.elm]: How do you extract message's payload with an if statement?

File: src/Platform.elm [hunk 6]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+Its first value is the initial state of your program's model.
+The second value is an initial command. Often that command is [Cmd.none](Platform-Cmd#none).
+
+### update
+
+The second argument, `update` is a function that takes these two arguments:
+
+1. A message from the outside world, that has the information your code will act upon.
+Usually this is a `Msg` type that you specify, (a constructor), which takes a record type
+matching the data coming from an incoming port. Unpack that data with `case` ... `of`.
```
Comment (Pilatch) [inline src/Platform.elm]: How about a link to here? https://www.elm-tutorial.org/en/02-elm-arch/03-messages.html

File: src/Platform.elm [hunk 7]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+
+2. The current state of your program's model, as maintained by the Elm runtime.
```
Comment (Pilatch) [inline src/Platform.elm]: I just deleted the back half of that sentence.

File: src/Platform.elm [hunk 8]
```diff
@@ -63,6 +63,85 @@ Initializing a headless program from JavaScript looks like this:
...
+
+The first argument, `init` is a [Tuple](Tuple) that lets you specify how your program should start.
+Its first value is the initial state of your program's model.
+The second value is an initial command. Often that command is [Cmd.none](Platform-Cmd#none).
+
+### update
+
+The second argument, `update` is a function that takes these two arguments:
+
+1. A message from the outside world, that has the information your code will act upon.
```
Comment (Pilatch) [inline src/Platform.elm]: Done.

File: src/Platform.elm (hunk 1 above)
Comment (Pilatch) [inline src/Platform.elm]: This does not appear to be true. I've been debugging a program here - https://github.com/Pilatch/elm-port-program-examples/blob/master/greeter/Greet.elm - and can't find any evidence to show that update is getting called multiple times when I supply a non-none command in the second half of the tuple that's returned from update.

Here I have the result of debugging the result of a Cmd.map:

    { type = "map", tagger = <function>, tree = { type = "leaf", home = "greet", value = "jomomma" } }

Here is the result of debugging Cmd.none:

    { type = "node", branches = [] }

They are significantly different, yet in both cases the update function is not run again, with the supplied message coming from Cmd.map.

File: src/Platform.elm (hunk 3 above)
Comment (Pilatch) [inline src/Platform.elm]: The code sample that follows that text illustrates that common use case, so I think it's safe to just delete the text in question.

Comment (Pilatch) [inline src/Platform.elm]: Thought about it. The common use case is important in the demystification process. Had I deleted text as I commented above, then my documentation would be no better than the type annotations.

Nevertheless I have taken the suggestion into account, reworded, and separated the common use case into a new paragraph.

File: src/Platform.elm (hunk 4 above)
Comment (Pilatch) [inline src/Platform.elm]: If I include that, then I have an init function, and may as well have a subscriptions function, then may as well have the Process.program call for completeness.

File: src/Platform.elm (hunk 5 above)
Comment (Pilatch) [inline src/Platform.elm]: Went the completeness route, and put one big code sample in there.

**Comment by Pilatch** [comment]
Regarding plumbing and tips for subscriptions, I'll have to look into when/how frequently subscriptions is called, then I'll better reform that section.

**Comment by evancz** [comment]
I want to keep the documentation as is. The `Platform.worker` function is not intended to be the first introduction to the `Program` type, so I think the best improvement may be to point folks to https://guide.elm-lang.org/ where they can learn the fundamentals. Having many places to learn the same thing becomes costly to maintain.

**Pull request closed without merging**